pkg/ebpf/uprobes/inspector.go

```diff
@@ -0,0 +1,176 @@
+	symbolMapBestEffort, _ := bininspect.GetAllSymbolsByName(elfFile, bestEffortSymbols)
+
+	funcMap := make(map[string]bininspect.FunctionMetadata, len(symbolMap)+len(symbolMapBestEffort))
+	for symbol, sym := range symbolMap {
+		m, err := p.symbolToFuncMetadata(elfFile, sym)
+		if err != nil {
+			return nil, false, fmt.Errorf("failed to convert symbol %v to function metadata: %w", sym, err)
+		}
+		funcMap[symbol] = *m
+	}
+	for symbol, sym := range symbolMapBestEffort {
+		m, err := p.symbolToFuncMetadata(elfFile, sym)
+		if err != nil {
+			return nil, false, fmt.Errorf("failed to convert symbol %v to function metadata: %w", sym, err)
+		}
+		funcMap[symbol] = *m
+	}
+
+	return funcMap, true, nil
+}
+
+func (*NativeBinaryInspector) symbolToFuncMetadata(elfFile *elf.File, sym elf.Symbol) (*bininspect.FunctionMetadata, error) {
+	manager.SanitizeUprobeAddresses(elfFile, []elf.Symbol{sym})
+	offset, err := bininspect.SymbolToOffset(elfFile, sym)
+	if err != nil {
+		return nil, err
+	}
+
+	return &bininspect.FunctionMetadata{EntryLocation: uint64(offset)}, nil
+}
+
+func (*NativeBinaryInspector) Cleanup(_ utils.FilePath) {
+	// Nothing to do here for the native inspector
+}
+
+// GoBinaryInspector is a BinaryInspector that inspects Go binaries, dealing with the specifics of Go binaries
+// such as the argument passing convention and the lack of uprobes
+type GoBinaryInspector struct {
+	structFieldsLookupFunctions map[bininspect.FieldIdentifier]bininspect.StructLookupFunction
+	paramLookupFunctions        map[string]bininspect.ParameterLookupFunction
+}
+
+// Ensure GoBinaryInspector implements BinaryInspector
+var _ BinaryInspector = &GoBinaryInspector{}
+
+// Inspect extracts the metadata required to attach to a Go binary from the ELF file at the given path.
+func (p *GoBinaryInspector) Inspect(fpath utils.FilePath, requests []SymbolRequest) (map[string]bininspect.FunctionMetadata, bool, error) {
+	path := fpath.HostPath
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, false, fmt.Errorf("could not open file %s, %w", path, err)
+	}
+	defer f.Close()
+
+	elfFile, err := elf.NewFile(f)
+	if err != nil {
+		return nil, false, fmt.Errorf("file %s could not be parsed as an ELF file: %w", path, err)
+	}
+
+	functionsConfig := make(map[string]bininspect.FunctionConfiguration, len(requests))
+	for _, req := range requests {
+		functionsConfig[req.Name] = bininspect.FunctionConfiguration{
+			IncludeReturnLocations: req.IncludeReturnLocations,
+			ParamLookupFunction:    p.paramLookupFunctions[req.Name],
+		}
+	}
+
+	inspectionResult, err := bininspect.InspectNewProcessBinary(elfFile, functionsConfig, p.structFieldsLookupFunctions)
+	if err != nil {
+		return nil, false, fmt.Errorf("error extracting inspection data from %s: %w", path, err)
+	}
+
+	return inspectionResult.Functions, true, nil
+}
+
+func (*GoBinaryInspector) Cleanup(_ utils.FilePath) {
+	// Nothing to do here for the native inspector
```

Comment:
```suggestion
	// Nothing to do here for the go binary inspector
```

Reply:
As both implementations (native and go) are empty, I'm wondering what do you imagine happening in the cleanup
If we don't have implementation, I'd suggest to remove the function until we have a use case

Reply:
It's for cleaning up the entries added to the params eBPF map. I'm still working on the refactor of USM code, not sure in which PR the Go binary inspector will end up.